### DISCUSSION TITLE
weird question

### DISCUSSION THREAD

**Discussion by atari-monk** [bilal-fazlani/commanddotnet#448] [Q&A]
Hi @drewburlingame 

I have weird question, but you are only guy i have contact to, i can even ask it.
I want to have database with list/info on my stuff, things i own.
I think i will do single table ef core code first.
Then i will copy cli app templete i worked out, with commandDotNet.
This shuld take one day, with test and first data in.
It seems best bet to make it work for me.
Is there faster solution i am missing ?
I dont want plain sql or excel, it seems to be layer i can avoid.

**Comment by drewburlingame** [comment]
Hi @krzm, it really depends so much on what your needs and constraints are.  It's not common for someone to use the console as their data entry point for the database, but it can make sense in some cases. If you use the console, making it intuitive for users is a lot of extra work, but maybe you don't need that if it's just for you.  Using aspnet minimal apis can also be a low effort way to build an application to take data.  It all depends on how you want to save and read the data. 

Have you tried asking questions on a site like stackoverflow?  you can use tags like below and there are people who are giddy with excitement to offer help for this kind of stuff.

- https://stackoverflow.com/questions/tagged/application-design
- https://stackoverflow.com/questions/tagged/system-design

Reddit can be another good resource

- https://www.reddit.com/r/dotnet/
- https://www.reddit.com/r/csharp/

I'd recommend posting across a few sites and see which ones give you the most useful feedback.

**Reply by atari-monk** [reply]
Thx @drewburlingame 
My need and constraint is, i want filled list of my stuff as fast and as simple as i can. 
It;s only for me.
I like console for this becouse it's a bit faster to make than gui.
I can also test entire app in integration test.
If i had users, yes i would use gui. 
Still prototype in console seems nice idea for testing.
I made Commander tutorial of Les jacson where he uses net web api and client like insomnia or postman to do this.
I know it's fast to. I dont know min api template yet.
Probablly it;s better than console becouse you can use it in docker and in the cloud somehow.
Yet i dont know how to make it awailable online. So it loses charm.
I invested some time in console and like to have it first.
Btw, i made to complex of a data model with to much many to many relations.
To much dictionaries, for every table and dictionary i made full acces and cli logic
also tables to print in console.
I got some data in this model and realized this is to costly to maintain.
I only need one main table and dictionaries can be seeded when db is created
or in managment studio (not in app).
I dont really want to post questions on this stuff.
This was/is just new exersice.

**Reply by drewburlingame** [reply]
One thing I would suggest is that you start with as few abstractions as possible in your design.  It's easy to overcomplicate it trying to follow a bunch of patterns that you don't end up needing.  Although, it can be a good learning exercise to see how it fails.  I'd recommend start with the bare minimum you need and then only add abstractions and extra layers when you need them.

As a bare minimum, for something very fast, you could have a DTO representing the table in the DB.  You could attribute the DTO with to use EF Core to save it to the DB.  You could also make it an IArgumentModel and decorate the properties as arguments.   Then your command is just saving the DTO to EF Core.   You could create an abstract class with Save and Read commands for a DTO and then subclass that for each DTO.

**Reply by atari-monk** [reply]
I been toying with cli apps for at least tow moths now.
I have build couple of helper libs or wrappers already. 
The standarize the way i do things.
Today i have db repo, sln, proj, made.
It uses ef-core-helper lib. 
It has generic repository form microsoft.doc, this pretty much setts all EF queries i need initially.
It provides configurable context. So i set two db's. One Local, second LocalTest. Just by setting json and usuing cli tools.
Also i have DI-Helper lib. It allows me to set up Dependency Incejtion by having sets where i register some module and suites where i register sets. Then i have also several libs to help me create apps.
For example i set config i json and my app registers components for integration test, or repls app, or command mode in CommandDotNet. It is pretty much a template at this point that i just copy between projects.
I am not willing to forgo this, becouse i will give some more time to copy this and minimal setup, for the ease of use it gives me.
I allways did this little tools. This time they are not so bad.
I have also my own lib for tables.
But i want to replace it with ones on github (there are two mainly)
I think this template is ok. It enables substraction of modules with config.
Problem was dbmodel. For app that needs to produce just data in form of a list, it was to complex.
Long story short it had many tables and all was with all commands in app. Even dictionaries.
When still prototyping table libs and integration test not done.
I made simpler model with one table, and dictionaires that are seeded just in dbContext.
I hope this will help to prototape corect db data fast, and also do integration test faster (one entity, prev. was many x many tests),
and also making only one console read table printout with new libs will be faster. 
I must finish this little proj this week.
I probably will.
I try to learn from tutorials, recently git-hub. (Not read one full lib yet) 
But I always do it my way anyway.
It;s just how mind works.

**Reply by atari-monk** [reply]
I suspect that looks insane
so maybe console is not the right proj here after all
well i'm still going with this
```
using System.ComponentModel.DataAnnotations;
using CommandDotNet;
using ModelHelper;

namespace Inventory.Min.Lib;

public class ItemInsertArgs 
    : Model
    , IArgumentModel
{
	[Operand(nameof(Name))
        , Required
        , MaxLength(NameMax)]
	public string? Name { get; set; }

    [Option('d', "description")
        , MaxLength(DescriptionMax)]
    public string? Description { get; set; }

	[Option('c', "categoryId")
        , Range(1, int.MaxValue, ErrorMessage = IdError)]
	public int? CategoryId { get; set; }

    [Option('p', "purchaseDate")]
	public DateTime? PurchaseDate { get; set; }

    [Option("currencyId")]
	public int? CurrencyId { get; set; } = 1;

    [Option('u', "purchasePrice")]
	public decimal? PurchasePrice { get; set; }

    [Option('s', "sellPrice")]
	public decimal? SellPrice { get; set; }

    [Option("imagePath")]
	public string? ImagePath { get; set; }

    [Option("lengthUnitId")]
	public int? LengthUnitId { get; set; } = 1;

    [Option('l', "length")]
	public double? Length { get; set; }

    [Option('h', "heigth")]
	public double? Heigth { get; set; }

    [Option("depth")]
	public double? Depth { get; set; }

    [Option("diameter")]
	public double? Diameter { get; set; }

    [Option("VolumeUnitId")
        , Range(1, int.MaxValue, ErrorMessage = IdError)]
	public int? VolumeUnitId { get; set; } = 1;

    [Option('v', "volume")]
	public double? Volume { get; set; }

    [Option('t', "TagId")
        , Range(1, int.MaxValue, ErrorMessage = IdError)]
	public int? TagId { get; set; }

    [Option('e', "StateId")
        , Range(1, int.MaxValue, ErrorMessage = IdError)]
	public int? StateId { get; set; }
}
```

**Reply by drewburlingame** [reply]
yeah, sometimes you've just got to try it to see what you think.  Listen, if it works for you, it's the right choice for you.  I think this would be more helpful if we had auto-complete for options, but we never finished that branch.

If you often have multiple entries at once, another option is to use a CSV file to collect the data and a command to parse the csv file.  Even more helpful if the data can be exported as CSV from another application.

**Comment by atari-monk** [comment]
I am chill testing to make sure it works before using runtime
```
public class ItemData
{
    private const string MainCmd = "item";
    private const string Cmd = "ins";
    private const string Name = "Name";
    private const string Description = "Description";

    public static IEnumerable<object[]> Insert01 =>
        new List<object[]>
        {
            new object[] 
            {
                0
                , new Item { Name = Name } 
                , new string[] { MainCmd, Cmd, Name }
            }
            ,
            new object[]
            {
                1
                , new Item { Name = Name, Description = Description } 
                , new string[] { MainCmd, Cmd, Name, "-d", Description }
            }
            ,
            new object[] 
            {
                2
                , new Item { Name = Name, CategoryId = 1 } 
                , new string[] { MainCmd, Cmd, Name, "-c", "1" }
            }
        };
}
```
```
using Inventory.Min.Cli.App.TestApi;
using Inventory.Min.Data;
using Xunit;
using XUnit.Helper;

namespace Inventory.Min.Cli.App.Tests;

[Collection(DbTests)]
[TestCaseOrderer(OrdererTypeName, OrdererAssemblyName)]
public class ItemInsertTests
    : OrderTest
        , IClassFixture<InventoryFixture>
{
    private InventoryFixture fixture;

    public ItemInsertTests(InventoryFixture fixture)
    {
        this.fixture = fixture;
    }

    [Theory]
    [MemberData(nameof(ItemData.Insert01)
        , MemberType= typeof(ItemData))]
    public void Test01(int index, Item expected, string[] cmd)
    {
        fixture.AssertItemCount(fixture.Uow, index);

        fixture.RunCmd(fixture.Booter, cmd);
        
        fixture.AssertItemCount(fixture.Uow, index + 1);
        
        var actual = fixture.GetItem(fixture.Uow, index);
        
        fixture.AssertItem(expected, actual);
    }
}
```

**Reply by drewburlingame** [reply]
If you've already created the console app, consider using https://commanddotnet.bilal-fazlani.com/testtools/harness/bdd/ to run the tests and that will exercise the entire console pipeline.  It will be like running the app from the console.

**Reply by atari-monk** [reply]
I am doing my tests by runnig whole app in fixture
tests are ordered and include database so i made api for testing
and tests in xunit
So far i dont test by output to console
but by get from database and comparing to expected

**Reply by atari-monk** [reply]
![image](https://user-images.githubusercontent.com/5488398/186749870-af156c68-6797-4e90-ae71-42187f0e1bcc.png)
I think app turned out ok. It works as i intended.

**Reply by drewburlingame** [reply]
nice work.  congratulations on your success.  🎉

**Reply by atari-monk** [reply]
Now i try to remake it in to web apps becouse most jobs are in this

**Comment by drewburlingame** [comment]
Hi @krzm 

We talked a while back about how to make a simpler cli for CRUD operations.  I saw you created a new one recently (it showed up as a in our list of github repos using CommandDotNet) and thought I'd throw together a quick suggestion on a way to structure the app that would be simpler, with less layers of indirection.   

With this app, you would just need an implementation of IRepository to register.  You could use a tool like AutoMapper or Mapster to auto-map the IArgumentModels to the classes you save to the database.

I think this will make your apps simpler to maintain and easier for others to grok when you get to the point of showing off your work.

```c#

public class Program
{
    public static void Main(string[] args)
    {
        IServiceProvider services = new ServiceCollection()
            .AddSingleton<Crud<Album>>()
            .AddSingleton<Crud<Song>>()
            .BuildServiceProvider();
        
        new AppRunner<Program>()
            .UseNameCasing(Case.KebabCase)
            .UseMicrosoftDependencyInjection(services)
            .Run(args);
    }

    [Subcommand(RenameAs = "albums")]
    public Crud<Album> Albums { get; set; }
    [Subcommand(RenameAs = "songs")]
    public Crud<Song> Songs { get; set; }
}

public class Crud<T> where T: IArgumentModel
{
    private readonly IRepository<T> _repository;

    public Crud(IRepository<T> repository) => 
        _repository = repository ?? throw new ArgumentNullException(nameof(repository));

    public void Create(IConsole console, T model)
    {
        console.WriteLine($"Creating {model}");
        _repository.Add(model);
    }
    
    public void Update(IConsole console, T model)
    {
        console.WriteLine($"Updating {model}");
        _repository.Update(model);
    }
    
    public void Delete(IConsole console, string name)
    {
        console.WriteLine($"Deleting {name}");
        _repository.Delete(name);
    }
    
    public void Read(IConsole console, string name)
    {
        console.WriteLine(_repository.Get(name));
    }
    
    public void List(IConsole console)
    {
        foreach (var model in _repository.GetAll())
        {
            console.WriteLine(model);
        }
    }
}

public class Album : IArgumentModel
{
    [Operand]
    public string Name { get; set; }

    public override string ToString()
    {
        return $"Album: {Name}";
    }
}

public class Song : IArgumentModel
{
    [Operand]
    public string Album { get; set; }
    [Operand]
    public string Name { get; set; }

    public override string ToString()
    {
        return $"Song: {Album} - {Name}";
    }
}

public class IRepository<T>
{
    public virtual IEnumerable<T> GetAll() => throw new NotImplementedException();
    public virtual T Get(string id) => throw new NotImplementedException();
    public virtual void Add(T entity) => throw new NotImplementedException();
    public virtual void Update(T entity) => throw new NotImplementedException();
    public virtual void Delete(string id) => throw new NotImplementedException();
}
```

Hope this is helpful.

Cheers,
Drew

**Reply by drewburlingame** [reply]
ps, this won't run until you've created and registered an implementation of IRepository.

**Reply by atari-monk** [reply]
Hi @drewburlingame 
Thx for template, i would like to test it in some new app.
I dont know witch of my repo you checked.
This is my latest [inventory-min-cli-app](https://github.com/krzm/inventory-min-cli-app)
You dont like the structure ?
I like to use same template, that i worked out, for all my cli app.
But maybe it can be much simplified as you suggest.
Right now i try to dokerize my web app, I struggle with that and dont know when this will end.
You know i may be terrible coder, since it is hard  for me to lose layers of indirection.
I managed to get free Azure Accont, so prolly I will be dokerizeing my web app.

**Reply by drewburlingame** [reply]
Hi @krzm,

It's not about whether I like the structure or not.  And it's not about whether you're a terrible coder.  Please don't take my suggestion as judgement of your skill.  I just remembered chatting before about layers of indirection.  I didn't have time when we chatted back then to put together an example of what I was talking about, but had a few moments the other day so thought I'd offer it.  I know you're still teaching yourself and thought it might be helpful to compare and contrast with what you have now.

**Reply by atari-monk** [reply]
Lets say i want to start new app. Simple notes taker. I want it to be simple as possible, but still plan ahead. First i need data layer.
I make separate repo, sln, proj for it (becouse i may use this db in many projects like api, or other apps). I use my Entity Framework Core library becouse i have abstracts for EF repository and unit of work there.
I will make nuget of this lib to simpliffy (used dlls before). I make my entity, context, migrations, repos, unit of work, also i make dtos witch i use for update (with automapper). I may add cli app to seed dictionaries. Now this is standard for me. At this point this seems sane to me and i dont see need to simpliffy nothing here. Would you agree ?

**Reply by drewburlingame** [reply]
This is certainly a common way to approach the problem, and if it's working for you, of course, there's nothing you need to change.  You've built a framework to wrap other frameworks. That's not an uncommon approach.  It has advantages and disadvantages.  I was just offering an alternative that could possibly be interesting and requires less wrapping of CommandDotNet.  It was interesting to me personally.  I liked using the generic type for the Subcommand as a way to keep the code strongly typed while not redefining the command class.

**Reply by atari-monk** [reply]
I cant go into it right now. I am a little under the weather with severe runny nose.
I have some fun with javasript right now.
https://krzm.github.io/javascript-game-prototype/
https://www.youtube.com/watch?v=5Vuvt4QAF48
I move towards web apps, api's and clouds.
I want to share my code with some users.
Need some feedback or fun or whatever with it.

**Reply by atari-monk** [reply]
I want a cli again. This time simple. No CRUD. So i will go with simplest template from your documentation page and it should be perfect.
